### DISCUSSION TITLE
TTGO Lora SX1278 - Unwanted Trailing bytes - FSK Mode - Transmit

### DISCUSSION THREAD

**Discussion by Pascal66** [jgromes/RadioLib#755] [Q&A]
Sending a frame ( a string or byte aarry)
and receveid this frame with rtl_433.
 I've some extra bytes at end, dont know why (I've disabled the CRC mode of the SX1278) 
```
time      : 2023-05-29 16:56:55
model     : fake       count     : 1             num_rows  : 1             rows      :
len       : 265          data      : 937c09004017e54159db70050143497004010043125c355f447285877dd91016 4a8
codes     : {265}937c09004017e54159db70050143497004010043125c355f447285877dd910164a8
Modulation: FSK          Freq1     : 433.5 MHz     Freq2     : 433.4 MHz
RSSI      : -0.1 dB      SNR       : 15.3 dB       Noise     : -15.4 dB
```
Theres always random bytes at end of the receveid frame
Originale is this one : `937c09004017e54159db70050143497004010043125c355f447285877dd91016`
Dont know if it a RadioLib problem, or a rtl_433 problem

Or for the string HelloWord ! 
`0c 48656c6c6f20576f726c6421 558` wich the good string : `012 Hello World! 5508`

Generaly, there's always 4a 4a8 4a0 550 55 558 trailing

I use : 
```
  uint8_t syncWord[] = {0x7f, 0xd9}; 
  state = radio.setSyncWord(syncWord, sizeof(syncWord));
  state = radio.setPreambleLength(32);
  state = radio.setDataShaping(RADIOLIB_SHAPING_NONE);
  state = radio.setEncoding(RADIOLIB_ENCODING_NRZ);
  state = radio.setCRC(false); 
  state = radio.fixedPacketLengthMode(sizeof(byteArr));
```

**Comment by jgromes** [comment]
I'm not all that familiar with rtl_433, but one thing to consider is - how does it know where the packet ends? SX1278 uses either a length byte that is in the packet itself (variable length mode) or packet length being known in advance (fixed length mode). How does rtl_433 know the length though? If it doesn't, there is no reason it should stop receiving data when the tranmission stops, it will just receive noise after the packet end - hence the random bytes.

**Reply by Pascal66** [reply]
That a good question... Length of received packet after preamble, or fixed length after preamble, or fixed length.

If it receive extra bytes in packet, it doesnt care about it and dont decode it.

How can rtl_433 catch any frame and decode it, but cant decode the same frame coming from the SX127x ?

But.. it cant be noise, as the trailing bytes a generaly the same and generaly a 55. It look like the SX127x doesnt stop after sending a frame, but prepare itself to send another one with a preamble.

Yes I do` finishTransmit(); `just after, or tested later. Same

**Reply by Pascal66** [reply]
An idea : 
Is the sending frame prefilled with 0x55 in an fixed byte array, before filling it with the user frame ?
I've to look deeper in radiolib

**Reply by jgromes** [reply]
> If it receive extra bytes in packet, it doesnt care about it and dont decode it.

Sorry, I don't understand. You seem to assume rtl_433 knows where the packet ends. I propose to examine this assumption in rtl_433.

> But.. it cant be noise, as the trailing bytes a generaly the same and generaly a 55

Possibly, or it can be the PA ramping down, like I said. Or it can be an unitialized array somewhere within rtl_433. Without further information it's just guesses.

> Yes I do finishTransmit();just after, or tested later. Same

Not sure what you are replying to - you haven't shared your full code, especially the part that does the actual transmission, so I don't know if it appropriate to do this or not.

> Is the sending frame prefilled with 0x55 in an fixed byte array

Not in RadioLib.

**Reply by Pascal66** [reply]
Ok. Thank you. Just asked.

Code is the same used by everyone trying to send a valid fsk frame.
No need to put it always each time.
Just the setting is important.

I just need a valid frame (Not a frame validated by another sx127x receiver)

For example :
I've a Fine Offset Electronics WH65B
I can received a valid frame with rtl_433
This weather station is too far from the receiver.
I want to repeat the valid(bis) frame with the ttgo lora
If the frame have some extra bytes on it, it can't be decoded nor by the receiver or anything else.

By valid, i mean a decoded frame

**Reply by jgromes** [reply]
> Code is the same used by everyone trying to send a valid fsk frame. No need to put it always each time. Just the setting is important.

I disagree - if I don't know which exact code you are using, I have no way to tell what might be wrong, so my answer will obviously be "from the provided information, nothing seems to be wrong" - if you want a better assesment, you will have to provide more information. I also don't know what "code used by everyone else" you're referring to. If you're referring to examples, those can change between different versions of the library.

> If the frame have some extra bytes on it, it can't be decoded nor by the receiver or anything else.

Again, how do you know there actually are extra data? And how do you know their presence prevents you or another receiver from decoding the frame? You can just disregard the data even if it is present and/or received, if the frame length is known in advance or encoded in the frame ...

**Reply by Pascal66** [reply]
Code is the most simple one in your example folder, called `SX172x_transmit.ino`

Instead of using `SX1278 radio = new Module(10, 2, 9, 3); `
i'm using`SX1278 radio = new Module(RADIO_CS_PIN, RADIO_DIO0_PIN, RADIO_RST_PIN, RADIO_BUSY_PIN); //Lilygo pins`

Outside setup(), i put `int state = radio.beginFSK();`
Instead of transmiting the simple "HelloWord! 
I transmit a byte[], with `state = radio.fixedPacketLengthMode(sizeof(byteArr));`

Inside setup(), I put :
```
  state = radio.setOOK(false);
  state = radio.packetMode();
  state = radio.setFrequency(433.44);
  state = radio.setBitRate(38.4);            // 100.0);
  state = radio.setFrequencyDeviation(19.2); // 10.0);
  state = radio.setRxBandwidth(250.0);
  state = radio.setOutputPower(10.0);
  state = radio.setCurrentLimit(60);
  state = radio.setRSSIConfig(2, 0);
  state = radio.setPreambleLength(32);
  uint8_t syncWord[] = {0x57, 0xfd, 0x99};
  state = radio.setSyncWord(syncWord, sizeof(syncWord));
  state = radio.setDataShaping(RADIOLIB_SHAPING_NONE);
  state = radio.setEncoding(RADIOLIB_ENCODING_NRZ);
  state = radio.setCRC(false); 

```
Nothing different than yours
I've tested all of combinations of those parameters, nothing more, nothing less

**Reply by Pascal66** [reply]
There's extra data, because, it's always the 0x55 preamble, wich is added to the end of the frame..
